File: iOS/sdk/Gypo/OrientationEvaluator.m
//
//  OrientationEvaluator.m
//  AliWeex
//
//  Created by 对象 on 2017/9/19.
//

#import "OrientationEvaluator.h"
#import "Vector3.h"
#import "Euler.h"

@interface OrientationEvaluator()

@property (nonatomic, assign) BOOL enabled;
@property (nonatomic, assign) int screenOrientation;

@property (nonatomic, strong) NSNumber * constraintAlpha;
@property (nonatomic, strong) NSNumber * constraintBeta;
@property (nonatomic, strong) NSNumber * constraintGamma;

@property (nonatomic, assign) double constraintAlphaOffset;
@property (nonatomic, assign) double constraintBetaOffset;
@property (nonatomic, assign) double constraintGammaOffset;

@property (nonatomic, strong) Quaternion *quaternion;
@property (nonatomic, strong) Quaternion *q0;
@property (nonatomic, strong) Quaternion *q1;
@property (nonatomic, strong) Vector3 *zee;
@property (nonatomic, strong) Euler *euler;

@end


@implementation OrientationEvaluator

- (instancetype)initWithConstraintAlpha:(NSNumber *)constraintAlpha constraintBeta:(NSNumber *)constraintBeta constraintGamma:(NSNumber *)constraintGamma {
    if (self = [self init]) {
        self.constraintAlpha = constraintAlpha;
        self.constraintBeta = constraintBeta;
        self.constraintGamma = constraintGamma;
        self.quaternion = [Quaternion quaternionWithX:0 y:0 z:0 w:1];
        self.zee = [Vector3 vectorWithX:0 y:0 z:1];
        self.q0 = [Quaternion quaternionWithX:0 y:0 z:0 w:0];
        self.q1 = [Quaternion quaternionWithX:sqrt(0.5) y:0 z:0 w:sqrt(0.5)];
        self.euler = [Euler new];
    }
    return self;
}

- (Quaternion *)calculateWithDeviceAlpha:(double)deviceAlpha deviceBeta:(double)deviceBeta  deviceGamma:(double)deviceGamma {
    double alpha = _constraintAlpha != nil ? [_constraintAlpha doubleValue] : (deviceAlpha + _constraintAlphaOffset);// Z
    double beta = _constraintBeta != nil ? [_constraintBeta doubleValue] : (deviceBeta + _constraintBetaOffset);// X
    double gamma = _constraintGamma != nil ? [_constraintGamma doubleValue] : (deviceGamma + _constraintGammaOffset);// Y
    
    // TODO 设备方向写死为0 纵向
    [self.euler setValueWithX:beta/180.0*M_PI y:alpha/180.0*M_PI z:-gamma/180.0*M_PI order:@"YXZ"];
    [_quaternion setFromEuler:self.euler];
    [_quaternion multiply:_q1 p:nil];
    [_quaternion multiply:[_q0 setFromAxisAngle:_zee angle:0] p:nil];
    
    return _quaternion;
}

@end

File: android/bindingx_weex_plugin/src/main/java/com/alibaba/android/binding/plugin/weex/internal/OrientationEvaluator.java
package com.alibaba.android.binding.plugin.weex.internal;

import android.support.annotation.Nullable;

/**
 * Description:
 *
 * Created by rowandjj(chuyi)<br/>
 */

class OrientationEvaluator {
    private Quaternion quaternion = new Quaternion(0,0,0,1);

    private Double constraintAlpha = null;
    private Double constraintBeta = null;
    private Double constraintGamma = null;

    private double constraintAlphaOffset = 0;
    private double constraintBetaOffset = 0;
    private double constraintGammaOffset = 0;


    OrientationEvaluator(@Nullable Double constraintAlpha, @Nullable Double constraintBeta, @Nullable Double constraintGamma) {
        this.constraintAlpha = constraintAlpha;
        this.constraintBeta = constraintBeta;
        this.constraintGamma = constraintGamma;
    }

    /**
     * @param deviceAlpha 陀螺仪alpha数据
     * @param deviceBeta 陀螺仪beta数据
     * @param deviceGamma 陀螺仪gamma数据
     * @param normalizedAlpha 校正后的alpha
     * */
    Quaternion calculate(double deviceAlpha, double deviceBeta, double deviceGamma, double normalizedAlpha) {
        double alpha = Math.toRadians(constraintAlpha != null ? constraintAlpha : (normalizedAlpha + constraintAlphaOffset));// Z
        double beta = Math.toRadians(constraintBeta != null ? constraintBeta : (deviceBeta + constraintBetaOffset));// X
        double gamma = Math.toRadians(constraintGamma != null ? constraintGamma : (deviceGamma + constraintGammaOffset));// Y

        // 设备方向写死为0 纵向
        setObjectQuaternion(quaternion, alpha, beta, gamma, 0);
        return quaternion;
    }


    private final Vector3 ZEE = new Vector3(0,0,1);
    private final Euler EULER = new Euler();
    private final Quaternion Q0 = new Quaternion();
    private final Quaternion Q1 = new Quaternion(-Math.sqrt(0.5),0,0,Math.sqrt(0.5));

    private void setObjectQuaternion(Quaternion quaternion, double alpha, double beta, double gamma, double orient) {
        EULER.setValue(beta,alpha,-gamma, "YXZ"); // 'ZXY' for the device, but 'YXZ' for us
        quaternion.setFromEuler(EULER); // orient the device
        quaternion.multiply(Q1); // camera looks out the back of the device, not the top
        quaternion.multiply(Q0.setFromAxisAngle(ZEE, -orient)); // adjust for screen orientation
    }

}
